ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java

```diff
@@ -196,67 +199,163 @@ public void run() {
   }
 }
 
+/**
+ * Tracks the arrival of events and allows waiting on all events of a particular type to arrive
+ */
 class Tracker {
   private final int numberOfReplicas;
-  private CountDownLatch totalReplicasCreated;
-  private CountDownLatch totalReplicasDeleted;
+  private final Helper creationHelper;
+  private final Helper deletionHelper;
+  private final ConcurrentMap<UpdateType, Helper> updateHelpers = new ConcurrentHashMap<>();
 
-  private ConcurrentHashMap<String, Boolean> creationSrcHosts = new ConcurrentHashMap<>();
-  private ConcurrentHashMap<String, Boolean> deletionSrcHosts = new ConcurrentHashMap<>();
+  /**
+   * Helper class that encapsulates the information needed to track a type of event
+   */
+  private class Helper {
+    private final ConcurrentHashMap<String, Boolean> hosts = new ConcurrentHashMap<>();
+    private final AtomicInteger notificationsReceived = new AtomicInteger(0);
+    private CountDownLatch latch = new CountDownLatch(numberOfReplicas);
+
+    /**
+     * Tracks the event that arrived on {@code host}:{@code port}.
+     * @param host the host that received the event
+     * @param port the port of the host that describes the instance along with {@code host}.
+     */
+    void track(String host, int port) {
+      notificationsReceived.incrementAndGet();
+      if (hosts.putIfAbsent(getKey(host, port), true) == null) {
+        latch.countDown();
+      }
+    }
+
+    /**
+     * Waits until the all replicas receive the event.
+     * @param duration the duration to wait for all the events to arrive
+     * @param timeUnit the time unit of {@code duration}
+     * @return {@code true} if events were received in all replicas within the {@code duration} specified.
+     * @throws InterruptedException
+     */
+    boolean await(long duration, TimeUnit timeUnit) throws InterruptedException {
+      return latch.await(duration, timeUnit);
+    }
+
+    /**
+     * Nullifies the notification received (if any) on the given {@code host}:{@code port}.
+     * @param host the host that to decrement on
+     * @param port the port of the host that describes the instance along with {@code host}.
+     */
+    void decrementCount(String host, int port) {
+      if (hosts.remove(getKey(host, port)) != null) {
+        long finalCount = latch.getCount() + 1;
```

Comment:
That's correct (on both counts) - this particular function is not thread safe and is actually not used in a concurrent way. It's a test helper (the test deletes the replica so it nullifies the notifications). It is used in a single thread and used after the test finishes awaits (i.e. there is no thread waiting when these functions are called)

Reply:
Looks like the class is only created within MockNotificationSystem and its methods invoked via synchronized methods anyway, so the first scenario won't happen (though the second one still could).  In lieu of re-engineering this thing (which may be overkill since it's use is limited to specific test code) we should at least document that this method is not thread safe and should not be run if other threads are waiting on await() (the second note should be extended to the methods in MockNotificationSystem that use it).

Another option could be to replace these decrementCount calls with a reset or clear method which resets the Tracker; when I was looking through ServerTestUtil it looked like a lot of the invocations were issued in groups that appeared to decrement all counts that were issued at that part of the test.

Reply:
Yeah since those changes should probably be done in a different PR, I'll leave a comment here saying that its not thread safe and should not be used while someone is on `await()` (including its users)